### ISSUE TITLE
Memory leak difference C and C++

### ISSUE THREAD

**Issue by wouter89** [facebook/infer#338]
Hi,

When analysing the following piece of code, Infer does report a memory leak when the file is saved as a C file and it reports no issues when the file is saved as a C++ file. I think Infer should also report a memory leak when this file is saved as C++. 

Does Infer make different assumptions for C++ compared to C files? 

```C
#include <stdlib.h>  //only needed for .cpp
#include <wchar.h>

void memoryLeakingFunction()
{
	char * data;
	data = (char *)calloc(100, sizeof(char));
}


int main()
{
	memoryLeakingFunction();
	return 0;
}
```

**Comment by ddino** [comment]
hi,

I saved this file as .c and then as .cpp and in both cases I get a leak.
infer -- clang -c blo.cpp 
Starting analysis (Infer version v0.8.0-a6d700d)
F..
Analyzed 1 file

Found 1 issue

blo.cpp:7: error: MEMORY_LEAK
   memory dynamically allocated to data by call to calloc() at line 7, column 18 is not reachable after line 7, column 3
  5.   {
  6.     char * data;
  7. >   data = (char *)calloc(100, sizeof(char));
  8.   }
  9.   
  10.   

Summary of the reports

  MEMORY_LEAK: 1

**Comment by akotulski** [comment]
I can't reproduce the problem locally, memory leak gets reported both in C and C++. It might be related to C headers in your system (they might #define calloc to be something else).  What platform are you using (mac os/linux distribution)?

**Comment by wouter89** [comment]
I ran the following commands on the file(s) mentioned above (one saved as c and one as cpp). This gave the following output:

------------
infer -- clang memleak1.c

Starting analysis (Infer version v0.5.0)
Computing dependencies... 100%
Analyzing 1 cluster. 100%
Analyzed 2 procedures in 1 file

Found 1 issue
memleak1.c:7: error: MEMORY_LEAK
   memory dynamically allocated to data by call to calloc() at line 7, column 17 is not reachable after line 7, column 2
  5.   {
  6.   	char * data;
  7. > 	data = (char *)calloc(100, sizeof(char));
  8.   }
  9.   

--------------

infer -- clang memleak1.cpp

Starting analysis (Infer version v0.5.0)
Computing dependencies... 100%
Analyzing 1 cluster. 100%
Analyzed 2 procedures in 1 file
No issues found

----------------

I'm running Infer version v0.5.0 (installed with homebrew recently) 
I'm running OS X El Capitan version 10.11.4
